deepspeed/comm/comm.py

```diff
@@ -697,8 +697,9 @@ def mpi_discovery(distributed_port=TORCH_DISTRIBUTED_DEFAULT_PORT, verbose=True)
 
     master_addr = None
     if rank == 0:
-        hostname_cmd = ["hostname -I"]
-        result = subprocess.check_output(hostname_cmd, shell=True)
+        import shlex
+        hostname_cmd = shlex.split("hostname -I")
```

Comment:
We may not need `shlex.split` when the command has no placeholder for injection. The same goes for some other fixes.

Reply:
For `hostname -I`, using `socket.gethostname` and `socket.gethostbyname_ex` can be safer.

Reply:
@tohtana, thanks for the feedback. I will remove the `shlex.split()` here. However, I am not getting the same output if I use `socket.gethostbyname_ex(socket.gethostname())`

Reply:
Actually, I think the `shlex.split()` is still need to pass args as a list since we are moving `shell=True`. Alternatively, I could manually construct the list. However, I think I will keep the `shlex.split()` to future-proof for arg changes.

Reply:
How about this code?  This is a general proposal to make the code robust against malicious modifications of system commands. But I don't think this is crucial because it won't be a typical attack that can harm our users. We can just keep `hostname -I` if this doesn't work.

```
>>> import socket
>>> hostname = socket.gethostname()
>>> ip_addresses = socket.gethostbyname_ex(hostname)[2]
>>> ip_addresses
['172.17.0.2']
```

Reply:
This is my vote, since it also helps resolve issues where we've seen `hostname -I` not available on all systems, such as Mac and latest inetutils on ArchLinux

Reply:
This is discrepancy that I get on my lambda box.  
```bash
>>> subprocess.check_output(["hostname", "-I"]).decode("utf-8").split()[0]
'192.168.1.104'
>>> socket.gethostbyname_ex(socket.gethostname())[2][0]
'127.0.1.1'
```

Maybe it is a problem with my setup. Can you guys test in your environment? I am happy to switch. Thanks.

Reply:
I see that as well, it looks like we would need to modify the command to be:

```
>>> socket.gethostbyname_ex(socket.gethostname()+".local")[2][0]
```

Thought it is slower to run on my system with that change.

Reply:
I think that is because `hostname -I` got the info from network interfaces like `eth0` while `socket` APIs relies on name resolving.
It's natural to expect DNS and/or `/etc/hosts` are properly set up on a multi-node run environment, but this definitely has an impact on many use cases.

Reply:
> I see that as well, it looks like we would need to modify the command to be:
> 
> ```
> >>> socket.gethostbyname_ex(socket.gethostname()+".local")[2][0]
> ```
> 
> Thought it is slower to run on my system with that change.

I wonder if `hosts` or DNS in your system (Windows?) has `[HOSTNAME].local` but it doesn't work on my env.

Reply:
> ```
> >>> socket.gethostbyname_ex(socket.gethostname()+".local")[2][0]
> ```
> 
This works and gives matching results on my Linux lambda box. 

@tohtana, do you mean this does not work for you?

Reply:
It doesn't work on my wsl.

```
$ python -c 'import socket; socket.gethostbyname_ex(socket.gethostname()+".local")[2][0]'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
socket.gaierror: [Errno -5] No address associated with hostname
```

Reply:
Interesting, I'd say lets leave it as hostname -I for now, and we can make another PR to update where we can more strenuously test Windows and other OSs?

Reply:
I agree to address this in another PR. This one is urgent and focuses on security.
But I'm not sure it is a good idea to get the IP from the first entry from `hostname -I`. It is not simple to control it even for the administrator. It is easier to tell users to configure `/etc/hosts` properly.
After a quick look on the usage, probably it can also be a hostname, not an IP.